precompiles/identity/src/lib.rs

```diff
@@ -0,0 +1,845 @@
+pub(crate) const SELECTOR_LOG_SUB_IDENTITY_ADDED: [u8; 32] =
+	keccak256!("SubIdentityAdded(address,address)");
+pub(crate) const SELECTOR_LOG_SUB_IDENTITY_REMOVED: [u8; 32] =
+	keccak256!("SubIdentityRemoved(address,address)");
+pub(crate) const SELECTOR_LOG_SUB_IDENTITY_REVOKED: [u8; 32] =
+	keccak256!("SubIdentityRevoked(address)");
+
+/// A precompile to wrap the functionality from pallet-identity
+pub struct IdentityPrecompile<Runtime>(PhantomData<Runtime>);
+
+#[precompile_utils::precompile]
+#[precompile::test_concrete_types(mock::Runtime)]
+impl<Runtime> IdentityPrecompile<Runtime>
+where
+	Runtime: pallet_identity::Config + pallet_evm::Config,
+	Runtime::AccountId: Into<H160>,
+	Runtime::Hash: From<H256>,
+	Runtime::RuntimeCall: Dispatchable<PostInfo = PostDispatchInfo> + GetDispatchInfo,
+	<Runtime::RuntimeCall as Dispatchable>::RuntimeOrigin: From<Option<Runtime::AccountId>>,
+	Runtime::RuntimeCall: From<pallet_identity::Call<Runtime>>,
+	BalanceOf<Runtime>: TryFrom<U256> + Into<U256> + solidity::Codec,
+{
+	// Note: addRegistrar(address) & killIdentity(address) are not supported since they use a
+	// force origin.
+
+	// editorconfig-checker-disable
+	#[precompile::public("setIdentity((((bool,bytes),(bool,bytes))[],(bool,bytes),(bool,bytes),(bool,bytes),(bool,bytes),(bool,bytes),bool,bytes,(bool,bytes),(bool,bytes)))")]
+	fn set_identity(
+		handle: &mut impl PrecompileHandle,
+		info: IdentityInfo<Runtime::MaxAdditionalFields>,
+	) -> EvmResult {
+		// editorconfig-checker-enable
+		let caller = handle.context().caller;
+
+		let event = log1(
+			handle.context().address,
+			SELECTOR_LOG_IDENTITY_SET,
+			solidity::encode_event_data(Address(caller)),
+		);
+		handle.record_log_costs(&[&event])?;
+
+		let call = pallet_identity::Call::<Runtime>::set_identity {
+			info: Self::identity_to_input(info)?,
+		};
+
+		let origin = Runtime::AddressMapping::into_account_id(caller);
+		RuntimeHelper::<Runtime>::try_dispatch(handle, Some(origin).into(), call)?;
+
+		event.record(handle)?;
+
+		Ok(())
+	}
+
+	#[precompile::public("setSubs((address,(bool,bytes))[])")]
+	fn set_subs(handle: &mut impl PrecompileHandle, subs: Vec<(Address, Data)>) -> EvmResult {
+		let mut call_subs = Vec::new();
+		for (i, (addr, data)) in subs.into_iter().enumerate() {
+			let addr = Runtime::AddressMapping::into_account_id(addr.into());
+			let data: pallet_identity::Data = data
+				.try_into()
+				.map_err(|e| RevertReason::custom(e).in_field(alloc::format!("index {i}")))?;
+			call_subs.push((addr, data));
+		}
+		let call = pallet_identity::Call::<Runtime>::set_subs { subs: call_subs };
+
+		let origin = Runtime::AddressMapping::into_account_id(handle.context().caller);
+		RuntimeHelper::<Runtime>::try_dispatch(handle, Some(origin).into(), call)?;
+
+		Ok(())
+	}
+
+	#[precompile::public("clearIdentity()")]
+	fn clear_identity(handle: &mut impl PrecompileHandle) -> EvmResult {
+		let caller = handle.context().caller;
```

Comment:
Do we need to do anything special here to make sure we weren't called from `delegatecall` or `staticcall`? (CC @nanocryk)

Reply:
Err, this should be handled by the way `PrecompileAt<_>` is set up.

Reply:
delegatecall is already forbidden by default by precompile-utils, as well as statuc calls if the fucntion not have the view modifier